core/modules/wbase/SendChannelShared.cc

```diff
@@ -0,0 +1,299 @@
+    scs->_keepAlive = scs;
+    return scs;
+}
+
+
+SendChannelShared::~SendChannelShared() {
+    if (!_sendChannel->isDead()) {
+        LOGS(_log, LOG_LVL_ERROR, "SendChannelShared destructor called while _sendChannel still alive.");
+        _sendChannel->kill();
+    }
+    if (_threadStarted == true) {
+        _lastRecvd = true;
+        _qFrontCv.notify_all();
+        _qBackCv.notify_all();
+    }
+}
+
+
+void SendChannelShared::setTaskCount(int taskCount) {
+    _taskCount = taskCount;
+}
+
+
+bool SendChannelShared::transmitTaskLast(StreamGuard sLock, bool inLast) {
+    /// _caller must have locked _streamMutex before calling this.
+    if (not inLast) return false; // This wasn't the last message buffer for this task, so it doesn't matter.
+    ++_lastCount;
+    bool lastTaskDone = _lastCount >= _taskCount;
+    return lastTaskDone;
+}
+
+
+bool SendChannelShared::kill(StreamGuard sLock) {
+    LOGS(_log, LOG_LVL_DEBUG, "SendChannelShared::kill() called");
+    bool ret = _sendChannel->kill();
+    _lastRecvd = true;
+    _qFrontCv.notify_all();
+    _qBackCv.notify_all();
+    return ret;
+}
+
+string SendChannelShared::makeIdStr(int qId, int jId) {
+    string str("QID" + (qId == 0 ? "" : to_string(qId) + "#" + to_string(jId)));
+    return str;
+}
+
+bool SendChannelShared::addTransmit(bool cancelled, bool erred, bool last, bool largeResult,
+                                    TransmitData::Ptr const& tData, int qId, int jId) {
+    QSERV_LOGCONTEXT_QUERY_JOB(qId, jId);
+    assert(tData != nullptr);
+    _run(); // Only starts the thread if it isn't already running.
+
+    /// Wait if there are enough TransmitData objects already in the queue.
+    bool reallyLast = _lastRecvd;
+
+    string idStr(makeIdStr(qId, jId));
+
+    if (reallyLast || isDead()) {
+        // If there's been some kind of error, make sure that nothing hangs waiting
+        // for this.
+        LOGS(_log, LOG_LVL_WARN, "addTransmit getting messages after isDead or reallyLast " << idStr);
+        _lastRecvd = true;
+        _qFrontCv.notify_all();
+        _qBackCv.notify_all();
+        return false;
+    }
+    {
+        lock_guard<mutex> streamLock(streamMutex);
+        reallyLast = transmitTaskLast(streamLock, last);
+    }
+
+    // Can this addTransmit wait?
+    // If erred or cancelled, this should be handled asap.
+    std::unique_lock<std::mutex> qLock(_queueMtx);
+    if (!erred && !cancelled) {
+        _qBackCv.wait(qLock, [this](){ return _transmitQueue.size() < 2; });
+    }
+    _transmitQueue.push(tData);
+    if (reallyLast || erred || cancelled) {
+        _lastRecvd = true;
+    }
+
+    qLock.unlock();
+    _qFrontCv.notify_one();
+    return true;
+}
+
+
+void SendChannelShared::_run() {
+    if (_threadStarted.exchange(true)) {
+        return;
+    }
+    if (isDead()) {
+        return;
+    }
+    std::thread thrd(&SendChannelShared::_transmitLoop, this);
+    _thread = std::move(thrd);
```

Comment:
This likely is explained somewhere but under what cases is the channel joinable vs not joinable? Usually it's one or the other by design.

Reply:
The unit test tend to segv when you don't join threads. 

It looks like the threads need to go away anyway. When the system is under load, they just sit there doing nothing, even though they have something to do, and the process runs out of threads. Between the threads and unsent messages, a bunch of memory is eaten up as well.